torch/nn/utils/parametrizations.py

```diff
@@ -1,10 +1,285 @@
+        if not hasattr(self, "base"):
+            # Note [right_inverse expm cayley]
+            # If we do not have use_trivialization=True, we just implement the inverse of the forward
+            # map for the Householder. To see why, think that for the Cayley map,
+            # we would need to find the matrix X \in R^{n x k} such that:
+            # Y = torch.cat([X.tril(), X.new_zeros(n, n - k).expand(*X.shape[:-2], -1, -1)], dim=-1)
+            # A = Y - Y.transpose(-2, -1).conj()
+            # cayley(A)[:, :k]
+            # gives the original tensor. It is not clear how to do this.
+            # Perhaps via some algebraic manipulation involving the QR like that of
+            # Corollary 2.2 in Edelman, Arias and Smith?
+            if self.orthogonal_map == OrthMaps.cayley or self.orthogonal_map == OrthMaps.matrix_exp:
+                raise NotImplementedError("It is not possible to assign to the matrix exponential "
+                                          "or the Cayley parametrizations when use_trivialization=False.")
+
+            # If parametrization == OrthMaps.householder, make Q orthogonal via the QR decomposition.
+            # Here Q is always real because we do not support householder and complex matrices.
+            # See note [Householder complex]
+            A, tau = torch.geqrf(Q)
+            # We want to have a decomposition X = QR with diag(R) > 0, as otherwise we could
+            # decompose an orthogonal matrix Q as Q = (-Q)@(-Id), which is a valid QR decomposition
+            # The diagonal of Q is the diagonal of R from the qr decomposition
+            A.diagonal(dim1=-2, dim2=-1).sign_()
+            # Equality with zero is ok because LAPACK returns exactly zero when it does not want
+            # to use a particular reflection
+            A.diagonal(dim1=-2, dim2=-1)[tau == 0.] *= -1
+            return A.transpose(-2, -1) if transpose else A
+        else:
+            if n == k:
+                # We check whether Q is orthogonal
+                if not _is_orthogonal(Q):
+                    Q = _make_orthogonal(Q)
+                else:  # Is orthogonal
+                    Q = Q.clone()
+            else:
+                # Complete Q into a full n x n orthogonal matrix
+                N = torch.randn(*(Q.size()[:-2] + (n, n - k)), dtype=Q.dtype, device=Q.device)
+                Q = torch.cat([Q, N], dim=-1)
+                Q = _make_orthogonal(Q)
+            self.base = Q
+
+            # It is necessary to return the -Id, as we use the diagonal for the
+            # Householder parametrization. Using -Id makes:
+            # householder(torch.zeros(m,n)) == torch.eye(m,n)
+            # Poor man's version of eye_like
+            neg_Id = torch.zeros_like(Q_init)
+            neg_Id.diagonal(dim1=-2, dim2=-1).fill_(-1.)
+            return neg_Id
+
+
+def orthogonal(module: Module,
+               name: str = 'weight',
+               orthogonal_map: Optional[str] = None,
```

Comment:
This is planned to be updated not be a str anymore?

Reply:
I think string is fine for now, as the `OrthMap`s will not be used anywhere else for now, and adding an enum increases the mental overhead for the user.

Reply:
I know we don't have that many precedents for that, but I really would like to see us moving away from string-based arguments. They are not easier to use, you always get some capitalization wrong, it doesn't play well with typing and you can't get auto-complete for it.
cc @mruberry what do you think?

Reply:
I second @albanD's opinion to prefer enums over strings for new APIs. Now that we don't support Python 2, I think there's not much reason to stick with string args

Reply:
@albanD raised the point that using enums helps typing these when you are using an IDE with autocomplete. Now, there's many people that do not work in this situation (me included).

For what is worth, in my opinion, strings are better as they are easier to remember, do not pollute the namespace, and create less mental overhead. I had a conversation with @rgommers the other day in which he also mentioned that strings are better than enums for APIs.

Reply:
I liked the idea of enums, until I actually tried using them for new APIs. The combination of more verbosity and API pollution is unlikely to be worth the minor benefits, unless you have a large number of options and/or the enum gets reused in multiple places (e.g., a set of padding modes which all functions that do padding understand).